### PR TITLE
MF1621 - Logical user removal

### LINKED ISSUE #1621: Add logical user removal for Mainflux users.
**FEATURE REQUEST**

1. Is there an open issue addressing this request? If it does, please add a "+1" reaction to the
   the existing issue, otherwise proceed to step 2.

2. Describe the feature you are requesting, as well as the possible use case(s) for it.

Add logical user removal for Mainflux users.

This means adding a flag for the user to determine whether it's active or not.
The flag is binary (true or false for two user states) but in the future, we might add different states.

The deactivated user can not log in.
When listing users, Admin can have a query parameter provided filter only inactive, only active, or all the users. By default, only active users are listed.

3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).

Should have

_In the future, all the policies attached to user should be removed as well, and reverting/activating user should be supported._

### PR DESCRIPTION
**Pull request by rodneyosodo**
Signed-off-by: 0x6f736f646f <blackd0t@protonmail.com>

### What does this do?
Remove the user logically. I have added a new field called active so when we remove the user we set the value to be false. Listing users, you are also given the opportunity to list removed used or active users

### Which issue(s) does this PR fix/relate to?
Resolves #1621

### Have you included tests for your changes?
N/A

### Did you document any new/modified functionality?
Not yet

### Notes
N/A

### CODE REVIEW

File: users/tracing/users.go
Comment (dborovcanin) [inline users/tracing/users.go]: I'm afraid bool is not enough to represent all the states: `active`, `inactive`, `all`.

File: users/users.go
Comment (dborovcanin) [inline users/users.go]: Maybe `Deactivate` is more suitable here.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: pkg/sdk/go/sdk.go
Comment (manuio) [inline pkg/sdk/go/sdk.go]: fix `existing thing`

File: pkg/sdk/go/sdk.go
Comment (manuio) [inline pkg/sdk/go/sdk.go]: Consider calling this method `DisableUser`

**Review by manuio — CHANGES_REQUESTED** [review]

File: pkg/sdk/go/sdk.go
Comment (rodneyosodo) [reply pkg/sdk/go/sdk.go]: I will change to `Deactivate` and `DeactivateUser` as indicated [here](https://github.com/mainflux/mainflux/pull/1620#discussion_r903986811)

File: users/tracing/users.go
Comment (rodneyosodo) [reply users/tracing/users.go]: Changed to using string

File: users/users.go
Comment (dborovcanin) [inline users/users.go]: I would suggest using `State` instead of `Active` here, with constraints (allowed values, case sensitivity) checked in the code. I would keep plain VARCHAR in the DB due to simplicity and readability (over enumeration and integer types).
Replace `active` with `state` at other places as well.

File: users/postgres/init.go
Comment (dborovcanin) [inline users/postgres/init.go]: Check if CHAR is the most suitable type here.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: users/postgres/users.go
Comment (dborovcanin) [inline users/postgres/users.go]: I'd prefer not to set default here but on the service layer with the rest of the service logic.

File: users/postgres/init.go
Comment (dborovcanin) [inline users/postgres/init.go]: Add a little more flexible size limit (64 should be plenty enough)

File: cli/users.go
Comment (dborovcanin) [inline cli/users.go]: Consider renaming `state` with `status`.

File: internal/apiutil/errors.go
Comment (dborovcanin) [inline internal/apiutil/errors.go]: Just `ErrInvalidStatus` here.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Consider enriching users' tests with a new user with an `inactive` status to see if tests pass.

File: users/mocks/users.go
Comment (dborovcanin) [inline users/mocks/users.go]: Please remove an empty line.

Comment (dborovcanin) [reply users/mocks/users.go]: There should be no need to sort users, but you can use `assert.ElementsMatch`.

File: users/service_test.go
Comment (dborovcanin) [inline users/service_test.go]: Add a case for logging in as a deactivated user.

File: pkg/sdk/go/sdk.go
Comment (dborovcanin) [inline pkg/sdk/go/sdk.go]: Consider renaming to something more generic like `ChangeUserStatus`.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: users/mocks/users.go
Comment (rodneyosodo) [reply users/mocks/users.go]: The `sortUsers` is needed when getting querying a list of users from allusers

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Rename `state` to `status`. It is a little less ambiguous.

File: cli/users.go
Comment (manuio) [inline cli/users.go]: I'd use `disable` instead of `deactivate`

File: users/api/metrics.go
Comment (manuio) [inline users/api/metrics.go]: This should be `change_user_status` but maybe better to find another naming. Why not `DisableUser` and `EnableUser`?

File: users/api/transport.go
Comment (manuio) [inline users/api/transport.go]: create a const for `active`

File: users/api/transport.go
Comment (manuio) [inline users/api/transport.go]: use it here

File: users/api/transport.go
Comment (manuio) [inline users/api/transport.go]: `"/users/:id"`

File: users/api/transport.go
Comment (manuio) [inline users/api/transport.go]: You could also fix this and use `id`

File: users/api/transport.go
Comment (manuio) [inline users/api/transport.go]: `id`

File: users/mocks/users.go
Comment (manuio) [inline users/mocks/users.go]: Probably better to create consts for `active` `inactive`

File: users/service.go
Comment (manuio) [inline users/service.go]: `state string, offset, limit uint64, email string, meta Metadat` should be passed as a `PageMetadata` structure

File: users/service.go
Comment (manuio) [inline users/service.go]: Same here

**Review by manuio — CHANGES_REQUESTED** [review]

File: users/api/metrics.go
Comment (dborovcanin) [reply users/api/metrics.go]: @0x6f736f646f Since we will have only these two states, and it's unlikely that will change in foreseeable future, the @manuio proposal sounds OK. It's a cleaner service interface and we can utilize the same repository implementation.

File: cli/users.go
Comment (drasko) [reply cli/users.go]: If the user's state is `active`, and not `enabled`, I'd leave `deactivate`.

Comment (manuio) [reply cli/users.go]: I'd change `active` by `enabled`. `active` could refer to the user activity in the platform.

File: cli/users.go
Comment (manuio) [inline cli/users.go]: `Change user status to enabled` ?

File: cli/users.go
Comment (manuio) [inline cli/users.go]: same here

File: internal/apiutil/errors.go
Comment (manuio) [inline internal/apiutil/errors.go]: Maybe better `invalid user account status`?

File: users/api/endpoint.go
Comment (manuio) [inline users/api/endpoint.go]: I'd use ` req.status`?

File: users/api/logging.go
Comment (manuio) [inline users/api/logging.go]: `status string`

File: users/api/logging.go
Comment (manuio) [inline users/api/logging.go]: `status`

File: users/api/requests.go
Comment (manuio) [inline users/api/requests.go]: `status`

**Review by manuio — CHANGES_REQUESTED** [review]

File: users/api/requests.go
Comment (manuio) [inline users/api/requests.go]: `status`

File: users/api/requests.go
Comment (manuio) [inline users/api/requests.go]: Create consts for these status.  Maybe even change them by `enabled and `disabled`?

**Review by manuio — CHANGES_REQUESTED** [review]

File: users/service.go
Comment (manuio) [inline users/service.go]: `EnabledStatusKey`

File: users/service.go
Comment (manuio) [inline users/service.go]: `DisabledSatusKey`

**Review by manuio — CHANGES_REQUESTED** [review]

File: cli/users.go
Comment (drasko) [reply cli/users.go]: Please leave "active"

Users cannot be enabled, they are not devices. Users are active/inactive and/or deactivated. Please keep this terminology, not enabled/disabled.

File: users/service.go
Comment (drasko) [reply users/service.go]: Users are active/inactive, not enabled.

File: users/service.go
Comment (drasko) [inline users/service.go:40]: Please do not use `enabled` terminology here, stick to `activated`. @dborovcanin please make sure the terminology is consistent, as I can see that there is some confusion which might lead to incosistence.

**Review by drasko — CHANGES_REQUESTED** [review]

File: cli/users.go
Comment (manuio) [reply cli/users.go]: Why a user can't be disabled? For example, Azure is using this therminology here: https://docs.microsoft.com/en-us/azure/active-directory/authentication/howto-mfa-userstates#azure-ad-multi-factor-authentication-user-states and there are many others...

Inactive can refer to a user not using the platform for a long time, right? In any case I agree to chose a naming and to be consistent in all the code.

File: users/service.go
Comment (manuio) [reply users/service.go:40]: @drasko I proposed this naming 3 days ago and @dborovcanin agreed. If you prefer active/inactive then @0x6f736f646f shoudl rename all

Comment (drasko) [reply users/service.go:40]: Active/inactive/deactivated.

File: users/tracing/users.go
Comment (mteodor) [reply users/tracing/users.go]: wouldnt enum be more suitable

File: users/api/endpoint.go
Comment (mteodor) [inline users/api/endpoint.go]: you can inline error check

File: users/service.go
Comment (mteodor) [inline users/service.go]: is this ok to be authentication error?

File: users/users.go
Comment (dborovcanin) [inline users/users.go]: While this is practically user update, no need to pass the entire user - the ID is sufficient.

File: users/tracing/users.go
Comment (dborovcanin) [inline users/tracing/users.go]: state -> status

File: users/api/requests.go
Comment (manuio) [inline users/api/requests.go]: Can you please split this cases:
```
	if req.status != users.AllStatusKey &&
	    req.status != users.EnabledStatusKey &&
	    req.status != users.DisabledStatusKey {
		return apiutil.ErrInvalidStatus
	}
```

File: users/api/requests.go
Comment (manuio) [inline users/api/requests.go]: same here

**Review by manuio — CHANGES_REQUESTED** [review]

File: users/service.go
Comment (dborovcanin) [inline users/service.go:362]: You can wrap both implementations of `EnableUser` and `DisableUser` with:
```go
func (svc usersService) changeStatus(ctx context.Context, token, id, status string) error {
	if _, err := svc.identify(ctx, token); err != nil {
		return err
	}

	dbUser, err := svc.users.RetrieveByID(ctx, id)
	if err != nil {
		return errors.Wrap(errors.ErrNotFound, err)
	}
	if dbUser.Status == status {
		return errStatus
	}

	return svc.users.ChangeStatus(ctx, dbUser, status)
}
```
and use it like:
```go
func (svc usersService) EnableUser(ctx context.Context, token, id string) error {
	if _, err := svc.changeStatus(ctx, token, id, EnabledStatusKey); err != nil {
		return errors.Wrap(ErrAlreadyEnabledUser, err)
	}
	return nil
}
```

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — DISMISSED** [review]

File: users/api/transport.go
Comment (manuio) [inline users/api/transport.go:126]: Do we need 2 endpoints to change the status? You could fusion them or even use update_user one, right?

Comment (manuio) [inline users/api/transport.go:126]: Do we need 2 endpoints to change the status? You could fusion them or even use update_user one, right? 

Don't forget to add the API doc here: https://github.com/mainflux/mainflux/blob/master/api/openapi/users.yml

**Review by manuio — CHANGES_REQUESTED** [review]

File: users/api/transport.go
Comment (dborovcanin) [reply users/api/transport.go:126]: We can, but I would say that this is OK at the moment. It's like an endpoint to enable and disable users - without exposing how it's done internally. If we add more status types, we definitely won't add an endpoint for each of them.

Comment (rodneyosodo) [reply users/api/transport.go:126]: I have updated the api documentation

File: api/openapi/users.yml
Comment (manuio) [inline api/openapi/users.yml]: `userId`

File: api/openapi/users.yml
Comment (manuio) [inline api/openapi/users.yml:228]: add section: 
```
      parameters:
        - $ref: "#/components/parameters/UserId"
```

File: api/openapi/users.yml
Comment (manuio) [inline api/openapi/users.yml]: same here

File: api/openapi/users.yml
Comment (manuio) [inline api/openapi/users.yml:248]: and here. BTW, you can verify with: https://editor.swagger.io/

**Review by manuio — CHANGES_REQUESTED** [review]

File: api/openapi/users.yml
Comment (rodneyosodo) [reply api/openapi/users.yml:248]: I didn't know about this. Thanks

File: api/openapi/users.yml
Comment (manuio) [inline api/openapi/users.yml]: `Disables a user account for a given user ID.`

**Review by manuio — CHANGES_REQUESTED** [review]

File: api/openapi/users.yml
Comment (manuio) [inline api/openapi/users.yml]: Sounds redundant. Maybe simply `User account status`?

**Review by manuio — CHANGES_REQUESTED** [review]

File: api/openapi/users.yml
Comment (manuio) [inline api/openapi/users.yml]: If I remember well the recommendation is to use pure PascalCase

Comment (rodneyosodo) [reply api/openapi/users.yml]: Yes

File: api/openapi/users.yml
Comment (manuio) [inline api/openapi/users.yml]: Do you mean `giving` or `for given`?

Comment (rodneyosodo) [reply api/openapi/users.yml]: `for a given`

File: api/openapi/users.yml
Comment (manuio) [inline api/openapi/users.yml]: There is an extra white space here `account  for`

File: api/openapi/users.yml
Comment (manuio) [inline api/openapi/users.yml]: `        default: enabled`

**Review by manuio — CHANGES_REQUESTED** [review]

File: api/openapi/users.yml
Comment (rodneyosodo) [reply api/openapi/users.yml]: Updated

**Review by manuio — DISMISSED** [review]
LGTM

**Comment by manuio** [comment]
@drasko Can we merge this one?

**Comment by drasko** [comment]
@manuio I'll review

**Comment by drasko** [comment]
This must wait for @dborovcanin's approval

**Comment by manuio** [comment]
@drasko He had approved before my last remark

File: users/postgres/init.go
Comment (dborovcanin) [inline users/postgres/init.go]: Consider using SMALLINT and enumeration for status. While it is more cryptic than VARCHAR, it's also unlikely that we'll have any other status anytime soon. Unlike simple boolean, it also leaves us the space for adding potential new statuses without changing the schema in the future.

Comment (rodneyosodo) [reply users/postgres/init.go]: `SMALLINT` and `ENUM` are both small and fast. `SMALLINT` have the advantage of changing to multiple values in the future without schema change but this needs to be defined somewhere so that someone can know the values for each int and are therefore more cryptic than `ENUM`.  `ENUM` has an additional advantage of being safe whereas in `VARCHAR` or `SMALLINT` you have to specify the check function. I propose the use of `ENUM` since I doubt that the values will change in the near future.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Since we now rely on DB enum, please also test repository behavior in case of inserting an invalid value for User status.

**Review by dborovcanin — APPROVED** [review]

**Review by drasko — APPROVED** [review]
LGTM

**Pull request merged**
